gnmi/gnmi.go

```diff
@@ -412,6 +417,78 @@ func (s *Server) Get(ctx context.Context, req *gpb.GetRequest) (*gpb.GetResponse
 	}
 }
 
+// PathAuth is an interface for checking authorization for gNMI paths.
+type PathAuth interface {
+	// Check returns if the user is allowed to read from or write from in the input path.
+	Check(path *gpb.Path, user string, write bool) bool
+}
+
+type subscribeWithAuth struct {
+	gpb.GNMI_SubscribeServer
+	auth PathAuth
+	user string
+}
+
+// Send implements gNMI subscribe send with authorization.
+func (s *subscribeWithAuth) Send(req *gpb.SubscribeResponse) error {
+	if req.GetSyncResponse() {
+		return s.GNMI_SubscribeServer.Send(req)
+	}
+	reqUpd := req.Response.(*gpb.SubscribeResponse_Update).Update
+
+	i := 0
+	for _, del := range reqUpd.Delete {
+		p, err := util.JoinPaths(reqUpd.GetPrefix(), del)
+		if err != nil {
+			return err
+		}
+		if s.auth.Check(p, s.user, false) {
+			reqUpd.Delete[i] = del
+			i++
+		}
+	}
+	reqUpd.Delete = reqUpd.Delete[:i]
+	i = 0
+	for _, upd := range reqUpd.Update {
+		p, err := util.JoinPaths(reqUpd.GetPrefix(), upd.GetPath())
+		if err != nil {
+			return err
+		}
+		if s.auth.Check(p, s.user, false) {
+			reqUpd.Update[i] = upd
+			i++
+		}
+	}
+	reqUpd.Update = reqUpd.Update[:i]
+	if len(reqUpd.Update) == 0 && len(reqUpd.Delete) == 0 {
+		return nil
+	}
+
+	return s.GNMI_SubscribeServer.Send(req)
+}
+
+// Subscribe wraps the internal subscribe with optional authorization.
+func (s *Server) Subscribe(srv gpb.GNMI_SubscribeServer) error {
+	p, _ := peer.FromContext(srv.Context())
+
+	if s.pathAuth == nil || p.Addr == nil {
```

Comment:
Hmm, when is the peer address nil?

Reply:
the reconcilers' LocalClient, added comment